File: langchain-typescript/rag.ts
// Import necessary modules
import * as fs from 'fs';
import { OpenAI } from "@langchain/openai";
import { TextLoader } from 'langchain/document_loaders/fs/text';
import { RecursiveCharacterTextSplitter } from 'langchain/text_splitter';
import { OpenAIEmbeddings } from "@langchain/openai";
import { FaissStore } from "@langchain/community/vectorstores/faiss";
import { RetrievalQAChain } from 'langchain/chains';
import { HoneyHiveLangChainTracer } from 'honeyhive';

// Async function to run the QA system
async function runQA() {
  const tracer = new HoneyHiveLangChainTracer({
    project: process.env.HH_PROJECT,
    sessionName: 'langchain-js-sotu-rag',
    apiKey: process.env.HH_API_KEY,
  });

  // Load the document
  const loader = new TextLoader('state_of_the_union.txt');
  const documents = await loader.load();

  // Split the document into chunks
  const textSplitter = new RecursiveCharacterTextSplitter({
    chunkSize: 1000,
    chunkOverlap: 200,
  });
  const docs = await textSplitter.splitDocuments(documents);

  // Create embeddings
  const embeddings = new OpenAIEmbeddings();

  // Create a FAISS vector store from the documents
  const vectorStore = await FaissStore.fromDocuments(docs, embeddings);

  // Create a retriever interface
  const retriever = vectorStore.asRetriever();

  // Initialize the OpenAI LLM
  const llm = new OpenAI({ temperature: 0 });

  // Create a RetrievalQA chain
  const qaChain = RetrievalQAChain.fromLLM(llm, retriever);

  // Ask a question
  const query = "What did the president say about Ketanji Brown Jackson?";
  const res = await qaChain.call({ query, callbacks: [tracer] });

  console.log(res.text);
}

// Run the QA system
runQA();

File: langchain-python/rag.py
import os
from langchain.document_loaders import TextLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.vectorstores import FAISS
from langchain.embeddings import OpenAIEmbeddings
from langchain.chains import RetrievalQA
from langchain.llms import OpenAI

# Load the document
loader = TextLoader('state_of_the_union.txt')
documents = loader.load()

# Split the document into chunks
text_splitter = RecursiveCharacterTextSplitter(chunk_size=1000, chunk_overlap=200)
docs = text_splitter.split_documents(documents)

# Create embeddings
embeddings = OpenAIEmbeddings()

# Create a FAISS vector store from the documents
vectorstore = FAISS.from_documents(docs, embeddings)

# Create a retriever interface
retriever = vectorstore.as_retriever()

# Initialize the OpenAI LLM
llm = OpenAI(temperature=0)

# Create a RetrievalQA chain
qa_chain = RetrievalQA.from_chain_type(
    llm=llm,
    chain_type="stuff",
    retriever=retriever
)

# Ask a question
query = "What did the president say about Ketanji Brown Jackson?"
result = qa_chain.run(query)

print(result)
